AWS (EC2+RDS)で構築したWordPressをterraform化する

naoya0408 opened the pull request:
プルリクエストしますので、レビューをお願いいたします。

terraformの実行結果
```
% terraform init
% terraform apply
```
<img width="932" alt="スクリーンショット 0001-11-03 18 29 48" src="https://user-images.githubusercontent.com/40429887/68083088-11312c80-fe68-11e9-870e-6f5cf7c8bf2e.png">

ansibleの実行結果
```
% ansible-playbook -i aws_ec2.yml playbook.yml                                                                                                                           (git)-[feature]

PLAY [wordpress] ************************************************************************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [nginx : enable amzn2extra-nginx1.12 repository] ***********************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [nginx : install nginx packages from amazon-linux-extras] **************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [nginx : start nginx] **************************************************************************************************************************************************************
changed: [web-server-1c]
changed: [web-server-1a]

TASK [nginx : copy nginx.conf] **********************************************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [php : amzn2-core.repo priority down] **********************************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [php : amzn2-extras.repo priority down] ********************************************************************************************************************************************
changed: [web-server-1c]
changed: [web-server-1a]

TASK [php : install epel/remi] **********************************************************************************************************************************************************
ok: [web-server-1a] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
ok: [web-server-1c] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
ok: [web-server-1c] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)
ok: [web-server-1a] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)

TASK [php : php72 install] **************************************************************************************************************************************************************
ok: [web-server-1c] => (item=php)
ok: [web-server-1a] => (item=php)
ok: [web-server-1c] => (item=php-devel)
ok: [web-server-1a] => (item=php-devel)
ok: [web-server-1c] => (item=php-fpm)
ok: [web-server-1a] => (item=php-fpm)
ok: [web-server-1c] => (item=php-gd)
ok: [web-server-1a] => (item=php-gd)
ok: [web-server-1c] => (item=php-mbstring)
ok: [web-server-1a] => (item=php-mbstring)
ok: [web-server-1c] => (item=php-mysqlnd)
ok: [web-server-1a] => (item=php-mysqlnd)

TASK [php : copy www.conf] **************************************************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [php : start php-fpm] **************************************************************************************************************************************************************
changed: [web-server-1a]
changed: [web-server-1c]

TASK [wordpress : download wordpress] ***************************************************************************************************************************************************
changed: [web-server-1a]
changed: [web-server-1c]

TASK [wordpress : create www directory] *************************************************************************************************************************************************
ok: [web-server-1c]
ok: [web-server-1a]

TASK [wordpress : unzip wordpress file] *************************************************************************************************************************************************
changed: [web-server-1c]
changed: [web-server-1a]

TASK [wordpress : permission change wordpress directory] ********************************************************************************************************************************
changed: [web-server-1c]
changed: [web-server-1a]

TASK [wordpress : restart nginx] ********************************************************************************************************************************************************
changed: [web-server-1c]
changed: [web-server-1a]

PLAY RECAP ******************************************************************************************************************************************************************************
web-server-1a              : ok=16   changed=7    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
web-server-1c              : ok=16   changed=7    unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

* パブリック IPにアクセスする→wordpressのセットアップを行う
<img width="1792" alt="スクリーンショット 0001-11-03 22 18 09" src="https://user-images.githubusercontent.com/40429887/68085788-0a1a1680-fe88-11e9-811d-53a5b204e95d.png">

* ロードバランサーの確認
<img width="1407" alt="スクリーンショット 0001-11-03 22 28 59" src="https://user-images.githubusercontent.com/40429887/68085908-5ade3f00-fe89-11e9-93ab-dc7cbf6a6d94.png">
ヘルスチェックはhealthyになっていることを確認

* ALBのDNS名を確認し、digコマンドで確認
```
% dig aws-wordpress-alb-*********.ap-northeast-1.elb.amazonaws.com

;; ANSWER SECTION:
aws-wordpress-alb-*********.ap-northeast-1.elb.amazonaws.com. 60 IN A 13.230.245.0
aws-wordpress-alb-*********.ap-northeast-1.elb.amazonaws.com. 60 IN A 13.115.25.219
```
アクセスが振り分けられていることがわかる

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

.gitignoreの新規追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

変数を定義するファイルの新規追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

ALBを作成する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

EC2で２つのインスタンスを作成する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

RDSでDBインスタンスを作成する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

使用するリージョンを定義する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

セキュリティーグループの作成

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

パブリック・プライベートに2つずつサブネットを作成する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

VPCの作成

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

インターネットゲートウェイの作成

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

ルートテーブルの作成

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

作成したルートテーブルをサブネットに関連付ける

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

ansibleを適用するターゲットの指定をする

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

ansibleのplaybookの新規追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

変数を定義する

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

nginxのhandlerを追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

nginxのtaskを追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

nginx.confを追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

phpのhandlerを追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

phpのtaskを追加

naoya0408 committed:
AWS (EC2+RDS)で構築したWordPressをterraform化する

www.confを追加

RVIRUS0817 commented on terraform-ansible/ansible/aws_ec2.yml:
```diff
@@ -0,0 +1,14 @@
+groups:
+  wordpress: "'web-server' in tags.Name"
+compose:
+  ansible_host: public_ip_address
```
この設定ってなんのためにあるのだろうか....
あとawsのクレデンシャルを設定していますが、基本セキュリティ上だめです。
やるならansible-vaultとか使って暗号化するといいです。

RVIRUS0817 commented on terraform-ansible/ansible/playbook.yml:
```diff
@@ -0,0 +1,12 @@
+---
+- hosts: wordpress
+  become: true
```
becomeはansible.cfgで以下のように書くと楽です。
```
[privilege_escalation]
become = True
```

RVIRUS0817 commented on terraform-ansible/ansible/playbook.yml:
```diff
@@ -0,0 +1,12 @@
+
+  vars:
+    - ansible_ssh_user: ec2-user
+    - ansible_ssh_private_key_file: ~/.ssh/aws_ssh_key.pem
```
この２つですが、ローカルまたはansibleサーバでデプロイするのであれば、sshまず最初に設定すると思うので、
必要ないと思いますよ。known_hostsに書かれますし。

RVIRUS0817 commented on terraform-ansible/ansible/roles/nginx/handlers/main.yaml:
```diff
@@ -0,0 +1,3 @@
+---
+- name: restart nginx
+  service: name=nginx state=restarted enabled=yes
```
`enabled=yes` はmain.ymlのときに起動して自動起動させるので必要ないですね。

RVIRUS0817 commented on terraform-ansible/ansible/roles/nginx/tasks/main.yaml:
```diff
@@ -0,0 +1,26 @@
+    name: nginx.service
+    state: restarted
+    daemon_reload: yes
+    enabled: yes
```
ここでやってるのでhandlerの部分いらない

RVIRUS0817 commented on terraform-ansible/ansible/roles/nginx/tasks/main.yaml:
```diff
@@ -0,0 +1,26 @@
+
+- name: copy nginx.conf
+  copy:
+    src: "../templates/nginx.conf"
```
`roles/nginx/teplates`
ですね

RVIRUS0817 commented on terraform-ansible/ansible/roles/php/tasks/main.yml:
```diff
@@ -0,0 +1,47 @@
+
+- name: copy www.conf
+  copy:
+    src: "../templates/www.conf"
```
ここも `roles/php/tempates/www.conf` ですね

RVIRUS0817 commented on terraform-ansible/ansible/roles/php/tasks/main.yml:
```diff
@@ -0,0 +1,47 @@
+    backup: no
+  notify: restart php-fpm
+
+- name: start php-fpm
+  systemd:
+    name: php-fpm.service
+    state: restarted
+    daemon_reload: yes
+    enabled: yes
```
ここはインストールしてから起動するので場所が初めの方にもっていきましょう。

RVIRUS0817 reviewed (commented):
@naoya0408 
コメントしました。terraformよく書けてます！
あとは細かいところ直してもらえばと！

naoya0408 committed:
aws_ec2.ymlの削除

naoya0408 committed:
hostsファイルの新規追加　EC2のパブリックIPを入力

naoya0408 committed:
ansible.cfgでsudo権限を付与する

naoya0408 committed:
handlers/main.ymlの削除

naoya0408 committed:
ファイルのコピー元のパスの変更

naoya0408 committed:
ファイルのコピー元のパスの変更

naoya0408 committed:
sshの設定の削除

naoya0408 commented:
## 変更点

* terraformにて作成したEC2のIPアドレスをhostsファイルに記述してansibleのターゲットを指定しました。
* 他レビューで指摘していただいた点の変更を行いました。

## ansibleの実行結果

```
% ansible-playbook -i hosts playbook.yml                                                                                                                                 (git)-[feature]

PLAY [all] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : enable amzn2extra-nginx1.12 repository] ***********************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : install nginx packages from amazon-linux-extras] **************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : start nginx] **************************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [nginx : copy nginx.conf] **********************************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [php : amzn2-core.repo priority down] **********************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [php : amzn2-extras.repo priority down] ********************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [php : install epel/remi] **********************************************************************************************************************************************************
changed: [52.198.31.60] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
changed: [3.112.222.233] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
ok: [52.198.31.60] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)
ok: [3.112.222.233] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)

TASK [php : php72 install] **************************************************************************************************************************************************************
changed: [52.198.31.60] => (item=php)
changed: [3.112.222.233] => (item=php)
changed: [52.198.31.60] => (item=php-devel)
changed: [3.112.222.233] => (item=php-devel)
changed: [52.198.31.60] => (item=php-fpm)
changed: [3.112.222.233] => (item=php-fpm)
changed: [52.198.31.60] => (item=php-gd)
changed: [3.112.222.233] => (item=php-gd)
changed: [52.198.31.60] => (item=php-mbstring)
changed: [3.112.222.233] => (item=php-mbstring)
changed: [52.198.31.60] => (item=php-mysqlnd)
changed: [3.112.222.233] => (item=php-mysqlnd)

TASK [php : start php-fpm] **************************************************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [php : copy www.conf] **************************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [wordpress : download wordpress] ***************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [wordpress : create www directory] *************************************************************************************************************************************************
ok: [3.112.222.233]
ok: [52.198.31.60]

TASK [wordpress : unzip wordpress file] *************************************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [wordpress : permission change wordpress directory] ********************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [wordpress : restart nginx] ********************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

RUNNING HANDLER [php : restart php-fpm] *************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

PLAY RECAP ******************************************************************************************************************************************************************************
3.112.222.233              : ok=17   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
52.198.31.60               : ok=17   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

naoya0408 committed:
作成したEC2のアドレスを指定

naoya0408 commented:
## 変更点

* terraformにて作成したEC2のIPアドレスをhostsファイルに記述してansibleのターゲットを指定しました。
* 他レビューで指摘していただいた点の変更を行いました。

## ansibleの実行結果

```
% ansible-playbook -i hosts playbook.yml                                                                                                                                 (git)-[feature]

PLAY [all] ******************************************************************************************************************************************************************************

TASK [Gathering Facts] ******************************************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : enable amzn2extra-nginx1.12 repository] ***********************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : install nginx packages from amazon-linux-extras] **************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [nginx : start nginx] **************************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [nginx : copy nginx.conf] **********************************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [php : amzn2-core.repo priority down] **********************************************************************************************************************************************
ok: [52.198.31.60]
ok: [3.112.222.233]

TASK [php : amzn2-extras.repo priority down] ********************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [php : install epel/remi] **********************************************************************************************************************************************************
changed: [52.198.31.60] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
changed: [3.112.222.233] => (item=http://rpms.famillecollet.com/enterprise/remi-release-7.rpm)
ok: [52.198.31.60] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)
ok: [3.112.222.233] => (item=https://dl.fedoraproject.org/pub/epel/epel-release-latest-7.noarch.rpm)

TASK [php : php72 install] **************************************************************************************************************************************************************
changed: [52.198.31.60] => (item=php)
changed: [3.112.222.233] => (item=php)
changed: [52.198.31.60] => (item=php-devel)
changed: [3.112.222.233] => (item=php-devel)
changed: [52.198.31.60] => (item=php-fpm)
changed: [3.112.222.233] => (item=php-fpm)
changed: [52.198.31.60] => (item=php-gd)
changed: [3.112.222.233] => (item=php-gd)
changed: [52.198.31.60] => (item=php-mbstring)
changed: [3.112.222.233] => (item=php-mbstring)
changed: [52.198.31.60] => (item=php-mysqlnd)
changed: [3.112.222.233] => (item=php-mysqlnd)

TASK [php : start php-fpm] **************************************************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [php : copy www.conf] **************************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [wordpress : download wordpress] ***************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

TASK [wordpress : create www directory] *************************************************************************************************************************************************
ok: [3.112.222.233]
ok: [52.198.31.60]

TASK [wordpress : unzip wordpress file] *************************************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [wordpress : permission change wordpress directory] ********************************************************************************************************************************
changed: [3.112.222.233]
changed: [52.198.31.60]

TASK [wordpress : restart nginx] ********************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

RUNNING HANDLER [php : restart php-fpm] *************************************************************************************************************************************************
changed: [52.198.31.60]
changed: [3.112.222.233]

PLAY RECAP ******************************************************************************************************************************************************************************
3.112.222.233              : ok=17   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
52.198.31.60               : ok=17   changed=11   unreachable=0    failed=0    skipped=0    rescued=0    ignored=0
```

## ブラウザで初期設定を行い、ロードバランサーのDNS名にアクセス
<img width="1792" alt="スクリーンショット 0001-11-17 0 37 51" src="https://user-images.githubusercontent.com/40429887/68995573-abff2180-08d2-11ea-85c3-6d8807c40a39.png">

RVIRUS0817 commented:
コード、テスト良いと思います！
